skbio/alignment/_tabular_msa.py

```diff
@@ -716,8 +730,119 @@ def reindex(self, key=None, keys=None):
             keys_.flags.writeable = False
 
         self._keys = keys_
+        self._minter = minter
+
+    def _fail_if_duplicate_keys(self, keys):
+        # Hashability of keys is implicitly checked here.
+        duplicates = find_duplicates(keys)
+        if duplicates:
+            raise UniqueError(
+                "Keys must be unique. Duplicate keys: %r" % duplicates)
 
     @experimental(as_of='0.4.0-dev')
+    def append(self, sequence, key=None, minter=None):
+        """Append a sequence to the MSA.
+
+        Parameters
+        ----------
+        sequence : alphabet-aware scikit-bio sequence object
+            Sequence to be appended. Must match the dtype of the MSA and the
+            number of positions in the MSA.
+        minter : callable or metadata key, optional
+            If None, the MSA is checked to see if a previous minter has been
+            cached. If not, an exception is raised. If a cached key is found
+            it is used to reindex the MSA.
+
+        Raises
+        ------
+        TypeError
+            If the sequence object is a type that doesn't have an alphabet
+        TypeError
+            If the type of the sequence does not match the dtype of the MSA.
+        ValueError
+            If the length of the sequence does not match the number of
+            positions in the MSA.
+        OperationError
+            If no key is provided, but the MSA has keys, and no cached key
+            is found.
+
+        See Also
+        --------
+        reindex
+
+        Notes
+        -----
+        The MSA is not automatically re-aligned when a sequence is appended.
+        Therefore, this operation is not necessarily meaningful on its own.
+
+        Examples
+        --------
+        >>> from skbio import DNA, TabularMSA
+        >>> msa = TabularMSA([DNA('')])
+        >>> msa.append(DNA(''))
+        >>> msa == TabularMSA([DNA(''), DNA('')])
+        True
+
+        >>> msa = TabularMSA([DNA('', metadata={'id': 'a'})], minter='id')
+        >>> msa.append(DNA('', metadata={'id': 'b'}), minter='id')
+        >>> msa == TabularMSA([DNA('', metadata={'id': 'a'}),
+        ...                    DNA('', metadata={'id': 'b'})], minter='id')
+        True
+        """
+
+        if key is not None and minter is not None:
```

Comment:
Hmmm... this is tricky. The current approach isn't too bad but there's an extra round of key validation happening that would be nice to get rid of. Here are a couple of ideas:

1. Modify `_assert_valid_keys` to validate keys, cast them to the correct type, and return them (this is currently happening in `reindex`, which is what the `keys` setter calls). In `append`, if the sequence and keys are valid, update `_keys` and `_seqs` directly (bypassing `reindex`/`keys` setter). Modify `reindex` to use `_assert_valid_keys` to avoid duplicate logic. `_assert_valid_keys` should be renamed since it is validating and casting.

2. In a try/except, append the sequence and use `keys` setter to set the new keys (which will also validate them). If any exception is raised, remove the appended sequence.

(1) is harder to implement but seems less of a hack than the current approach and (2).

(2) is easy to implement but feels very much like a hack, similar to the current implementation. (2) will be more efficient than the current approach.

What do you think? @ebolyen can you weigh in here?